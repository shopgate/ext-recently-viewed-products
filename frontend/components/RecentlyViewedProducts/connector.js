import { connect } from 'react-redux';
import { getRecentlyViewedProducts } from '../../selectors';
import { RECENTLY_VIEWED_PRODUCTS_SLIDER_LIMIT } from '../../constants';

/**
 * Maps the contents of the state to the component props.
 * @param {Object} state The current application state.
 * @param {Object} props The component properties.
 * @return {Object} The extended component props.
 */
const mapStateToProps = state => ({
  products: getRecentlyViewedProducts(state, RECENTLY_VIEWED_PRODUCTS_SLIDER_LIMIT),
});

export default connect(mapStateToProps);
