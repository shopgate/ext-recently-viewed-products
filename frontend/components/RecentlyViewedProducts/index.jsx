import React from 'react';
import PropTypes from 'prop-types';
import I18n from '@shopgate/pwa-common/components/I18n';
import Slider from '@shopgate/pwa-common/components/Slider';
import Card from '../Card';
import ProductCard from '../ProductCard';
import connect from './connector';
import styles from './style';

/**
 * Creates an item for a single product.
 * @param {Object} product The product data.
 * @return {JSX} The rendered product card.
 */
const createSliderItem = (product ) => {
  const key = `s${product.id}`;
  return (
    <Slider.Item key={key} className={styles.sliderItem}>
      <Card className={styles.card}>
        <ProductCard
          product={product}
          titleRows={2}
        />
      </Card>
    </Slider.Item>
  );
};

/**
 * The RecentlyViewedProducts component
 * @return {JSX}
 */
const RecentlyViewedProducts = ({ products, settings }) => {
  const { sliderSettings } = settings;

  const items = products.map((
    product => createSliderItem(product, settings)
  ));

  if (!items.length) {
    return null;
  }

  // Finally, build the slider.
  return (
    <div className={styles.slider}>
      <h3 className={styles.headline}>
        <I18n.Text string="recently_viewed_products.headline" />
      </h3>
      <Slider
        autoPlay={sliderSettings.autoPlay}
        loop={false}
        indicators={false}
        controls={false}
        interval={Number.parseInt(sliderSettings.delay, 10)}
        snapItems={false}
        slidesPerView={2.3}
        classNames={{ container: styles.sliderContainer }}
      >
        {items}
      </Slider>
    </div>
  );
};

RecentlyViewedProducts.propTypes = {
  products: PropTypes.arrayOf(PropTypes.shape()),
  // The settings as received by the pipeline request
  settings: PropTypes.shape({
    headline: PropTypes.string,
    sliderSettings: PropTypes.shape({
      autostart: PropTypes.bool,
      delay: PropTypes.oneOfType([ // The delay between the automatic slides loops
        PropTypes.string,
        PropTypes.number,
      ]),
    }),
    showName: PropTypes.bool,
  }),
};

RecentlyViewedProducts.defaultProps = {
  settings: {
    headline: 'Recently viewed',
    sliderSettings: {
      autostart: false,
      delay: null,
    },
    showName: false,
  },
  products: [],
};

export default connect(RecentlyViewedProducts);
