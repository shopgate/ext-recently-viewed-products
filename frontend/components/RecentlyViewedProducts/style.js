import { css } from 'glamor';
import { themeConfig } from '@shopgate/pwa-common/helpers/config';

const { colors, variables } = themeConfig;

const sliderContainer = css({
  marginLeft: 'auto',
  marginRight: 'auto',
  position: 'relative',
  paddingBottom: variables.gap.small,
}).toString();

const slider = css({
  width: '100%',
  flex: 1,
  padding: variables.gap.small,
}).toString();

const sliderItem = css({
  width: '50%',
}).toString();

const productInfo = css({
  padding: 15,
}).toString();

const productName = css({
  fontWeight: 500,
}).toString();

const priceGrid = css({
  alignItems: 'flex-end',
}).toString();

const priceStrikedItem = css({
  flexGrow: 1,
}).toString();

const priceItem = css({
  flexGrow: 1,
  textAlign: 'right',
}).toString();

const priceBase = css({
  padding: '0 15px',
}).toString();

const priceStriked = css({
  ...priceBase,
  fontSize: '0.875rem',
}).toString();

const price = css({
  ...priceBase,
  color: colors.primary,
  fontSize: '1rem',
}).toString();

const card = css({
  background: '#FFF',
  height: '100%',
  margin: '0px 8px',
}).toString();

const headline = css({
  fontSize: 18,
  margin: `0 0 ${variables.gap.big}px`,
  textAlign: 'center',
}).toString();

const headlineCart = css({
  fontSize: 'medium',
  margin: `${variables.gap.small}px`,
  textAlign: 'left',
}).toString();

const showMoreContainer = css({
  display: 'flex',
  justifyContent: 'flex-start',
  textAlign: 'left',
  marginBottom: -variables.gap.bigger,
});

export default {
  card,
  headline,
  headlineCart,
  sliderContainer,
  slider,
  sliderItem,
  productInfo,
  productName,
  priceGrid,
  priceStrikedItem,
  priceItem,
  priceBase,
  priceStriked,
  price,
  showMoreContainer,
};
