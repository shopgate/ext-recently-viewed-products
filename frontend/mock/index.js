import configureStore from 'redux-mock-store';
import { QUICKLINKS_MENU } from '@shopgate/pwa-common/constants/MenuIDs';
import { REDUX_NAMESPACE_RECENTLY_VIEWED_PRODUCTS } from '../constants';
import { pageId } from '../config';

export const defaultState = {
  extensions: {
    [REDUX_NAMESPACE_RECENTLY_VIEWED_PRODUCTS]: {
      isFetching: false,
      productIds: [],
    },
  },
  product: {
    currentProduct: { productId: null },
  },
  history: { pathname: '' },
  menu: {
    menusById: {

    },
  },
};

/**
 * Creates a mocked store with the passed amount of products.
 * @param {number} [amount=2] The amount of products.
 * @param {boolean} [createPageEntry=true] Tells if an entry for a recently viewed products page
 *   shall be added.
 * @return {Object}
 */
export const createStateWithProducts = (amount = 2, createPageEntry = true) => {
  const productIds = [];
  const productsById = {};

  for (let i = 1; i <= amount; i += 1) {
    productIds.push(i.toString());
    productsById[i] = {
      productData: {
        id: i.toString(),
        featuredImageUrl: `https://example.com/${i}`,
        name: `${i}`,
      },
    };
  }

  return {
    product: {
      productsById,
    },
    extensions: {
      [REDUX_NAMESPACE_RECENTLY_VIEWED_PRODUCTS]: {
        isFetching: false,
        productIds,
      },
    },
    menu: {
      menusById: {
        [QUICKLINKS_MENU]: {
          isFetching: false,
          entries: [{
            url: `/page/${createPageEntry ? pageId : 'noop'}`,
            label: 'Recently viewed',
          }],
        },
      },
    },
  };
};

export const stateWithoutValidMenuEntries = {
  menu: {
    menusById: {
      [QUICKLINKS_MENU]: {
        isFetching: false,
        entries: [{
          url: '/page/foobar',
          label: 'Recently viewed',
        }],
      },
    },
  },
};

/**
 * Empty store.
 * @returns {function}
 */
export const getEmptyStore = () => configureStore()(defaultState);

// eslint-disable-next-line require-jsdoc
export const getStoreWithProducts = (amount = 2, createPageEntry = true) =>
  configureStore()(createStateWithProducts(amount, createPageEntry));
