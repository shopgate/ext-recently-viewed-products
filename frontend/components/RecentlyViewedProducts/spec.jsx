import React, { Component as mockedComponent } from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { getEmptyStore, getStoreWithProducts } from '../../mock';
import RecentlyViewedProducts from './index';

/* eslint-disable require-jsdoc, global-require */
jest.mock('@shopgate/pwa-common/components/Slider', () => (class extends mockedComponent {
  static Item({ children }) { return (<div>{children}</div>); }
  render() {
    return (<div>{this.props.children}</div>);
  }
}));

jest.mock('@shopgate/pwa-common-commerce/market/helpers/showTaxDisclaimer', () => () => null);

jest.mock(
  '@shopgate/pwa-common/helpers/config',
  () => require('@shopgate/pwa-common/helpers/config/mock')
);
/* eslint-enable */

describe('RecentlyViewedProducts', () => {
  it('should render nothing', () => {
    const component = mount(
      <Provider store={getEmptyStore()}>
        <RecentlyViewedProducts />
      </Provider>
    );
    expect(component.html()).toBe(null);
  });
  it('should render slider', () => {
    const component = mount(
      <Provider store={getStoreWithProducts()}>
        <RecentlyViewedProducts />
      </Provider>
    );
    expect(component.find('Item').length).toBe(2);
    expect(component).toMatchSnapshot();
  });
});
