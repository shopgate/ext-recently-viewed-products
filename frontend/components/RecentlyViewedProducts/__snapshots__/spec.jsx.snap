// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecentlyViewedProducts should render slider 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(RecentlyViewedProducts)>
    <RecentlyViewedProducts
      dispatch={[Function]}
      products={
        Array [
          Object {
            "featuredImageUrl": "https://example.com/one",
            "id": 1,
            "name": "one",
          },
          Object {
            "featuredImageUrl": "https://example.com/two",
            "id": 2,
            "name": "two",
          },
        ]
      }
      settings={
        Object {
          "headline": "Recently viewed",
          "showName": false,
          "sliderSettings": Object {
            "autostart": false,
            "delay": null,
          },
        }
      }
    >
      <div
        className="css-1qazn57"
      >
        <h3
          className="css-rsku8h"
        >
          <Translate
            className=""
            params={Object {}}
            string="recently_viewed_products.headline"
          >
            <span
              className=""
            >
              recently_viewed_products.headline
            </span>
          </Translate>
        </h3>
        <_class
          classNames={
            Object {
              "container": "css-y5yof0",
            }
          }
          controls={false}
          indicators={false}
          interval={NaN}
          loop={false}
          slidesPerView={2.3}
          snapItems={false}
        >
          <div>
            <Item
              className="css-j85wo1"
              key="s1"
            >
              <div>
                <Card
                  className="css-6r0d5j"
                >
                  <div
                    className="css-1l0if7f css-6r0d5j"
                  >
                    <ProductCard
                      product={
                        Object {
                          "featuredImageUrl": "https://example.com/one",
                          "id": 1,
                          "name": "one",
                        }
                      }
                      titleRows={2}
                    >
                      <Link
                        className=""
                        href="/item/31"
                        itemProp="item"
                        itemScope={true}
                        itemType="http://schema.org/Product"
                        tagName="a"
                      >
                        <div
                          aria-hidden={true}
                          className="css-c6ki83 "
                          onClick={[Function]}
                          role="link"
                        >
                          <ProductImage
                            alt="one"
                            animating={true}
                            forcePlaceholder={false}
                            highestResolutionLoaded={[Function]}
                            itemProp="image"
                            ratio={null}
                            resolutions={
                              Array [
                                Object {
                                  "blur": 2,
                                  "height": 50,
                                  "width": 50,
                                },
                                Object {
                                  "height": 440,
                                  "width": 440,
                                },
                              ]
                            }
                            src="https://example.com/one"
                          >
                            <Image
                              alt="one"
                              animating={true}
                              backgroundColor="#f2f2f2"
                              className=""
                              forcePlaceholder={false}
                              highestResolutionLoaded={[Function]}
                              itemProp="image"
                              onError={[Function]}
                              onLoad={null}
                              ratio={null}
                              resolutions={
                                Array [
                                  Object {
                                    "blur": 2,
                                    "height": 50,
                                    "width": 50,
                                  },
                                  Object {
                                    "height": 440,
                                    "width": 440,
                                  },
                                ]
                              }
                              src="https://example.com/one"
                              transition={null}
                            >
                              <div
                                className="css-326p2s "
                              />
                            </Image>
                          </ProductImage>
                        </div>
                      </Link>
                    </ProductCard>
                  </div>
                </Card>
              </div>
            </Item>
            <Item
              className="css-j85wo1"
              key="s2"
            >
              <div>
                <Card
                  className="css-6r0d5j"
                >
                  <div
                    className="css-1l0if7f css-6r0d5j"
                  >
                    <ProductCard
                      product={
                        Object {
                          "featuredImageUrl": "https://example.com/two",
                          "id": 2,
                          "name": "two",
                        }
                      }
                      titleRows={2}
                    >
                      <Link
                        className=""
                        href="/item/32"
                        itemProp="item"
                        itemScope={true}
                        itemType="http://schema.org/Product"
                        tagName="a"
                      >
                        <div
                          aria-hidden={true}
                          className="css-c6ki83 "
                          onClick={[Function]}
                          role="link"
                        >
                          <ProductImage
                            alt="two"
                            animating={true}
                            forcePlaceholder={false}
                            highestResolutionLoaded={[Function]}
                            itemProp="image"
                            ratio={null}
                            resolutions={
                              Array [
                                Object {
                                  "blur": 2,
                                  "height": 50,
                                  "width": 50,
                                },
                                Object {
                                  "height": 440,
                                  "width": 440,
                                },
                              ]
                            }
                            src="https://example.com/two"
                          >
                            <Image
                              alt="two"
                              animating={true}
                              backgroundColor="#f2f2f2"
                              className=""
                              forcePlaceholder={false}
                              highestResolutionLoaded={[Function]}
                              itemProp="image"
                              onError={[Function]}
                              onLoad={null}
                              ratio={null}
                              resolutions={
                                Array [
                                  Object {
                                    "blur": 2,
                                    "height": 50,
                                    "width": 50,
                                  },
                                  Object {
                                    "height": 440,
                                    "width": 440,
                                  },
                                ]
                              }
                              src="https://example.com/two"
                              transition={null}
                            >
                              <div
                                className="css-326p2s "
                              />
                            </Image>
                          </ProductImage>
                        </div>
                      </Link>
                    </ProductCard>
                  </div>
                </Card>
              </div>
            </Item>
          </div>
        </_class>
      </div>
    </RecentlyViewedProducts>
  </Connect(RecentlyViewedProducts)>
</Provider>
`;
